### PR TITLE
Configuration file for Commands and API Keys

### PR DESCRIPTION
**Pull request by TerryAg**
Added a new configuration file (config.ini) which keeps track of what commands are enabled/disabled, as well as storing API keys so that they don't need to be environment variables anymore.

config/default.ini is the default format of the configuration file, but you'll need to rename it to config/config.ini and add your API keys there (config/config.ini is in the gitignore).

Commands can simply be disabled by swapping "on" with "off". For each new command a new row will have to be added, where the key is the same as the class' name.

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -4,14 +4,15 @@
...
-import json, os, sys
+import json, os, sys, configparser
 
 import discord
 import asyncio
 
 import commands
 
 client = discord.Client()
+config = configparser.ConfigParser()
```
Comment (VineBalloon) [inline bot.py]: Should move all config parsing logic here.

File: bot.py [hunk 2]
```diff
@@ -58,12 +59,17 @@
             args = '\\n '.join(message.content[1:].splitlines()).split()
             if args:
                 cls, args = commands.Helpme.find_command(args)
-                output = await cls(client, message).init(*args)
-                if isinstance(output, discord.Embed):
-                    await client.send_message(message.channel, embed=output)
-                elif output is not None:
-                    await client.send_message(message.channel, output)
+                if config['COMMANDS'].getboolean(cls.__name__):
+                    # Command is not disabled
```
Comment (VineBalloon) [inline bot.py]: Avoid double negatives

File: bot.py [hunk 3]
```diff
@@ -58,12 +59,17 @@
...
+                    elif output is not None:
+                        await client.send_message(message.channel, output)
     except discord.errors.HTTPException as e:
         await client.send_message(message.channel, err)
 
-client.run(os.environ['TOKEN'])
+if config.read('config/config.ini'):
+    client.run(config['KEYS']['DiscordToken'])
+else:
+    print("Can't find config file! Have you added your keys to the default config file and renamed the file to 'config.ini'?")
```
Comment (VineBalloon) [inline bot.py]: Should raise exception instead, bot shouldn't run without a config file (we provide one in the repo). 
You may want to use `read_file()` and `open()` for this instead and extend the `FileNotFound` exception thrown by `open()` to something like `ConfigFileNotFound` with a helpful error message.

File: commands/music.py [hunk 4]
```diff
@@ -17,7 +17,9 @@
 # Please ensure that you have enabled the YouTube Data API for your project.
 
 # Make sure you set your environment variable
-DEVELOPER_KEY = os.environ['YT_API']
+config = configparser.ConfigParser()
+config.read('config/config.ini')
```
Comment (VineBalloon) [inline commands/music.py]: Don't read multiple times, make an import like `from bot import config` and use that wherever the config is required.

File: config/default.ini [hunk 5]
```diff
@@ -0,0 +1,23 @@
+[KEYS]
+YouTube=
+DiscordToken=
+TwitchClientID=
+
```
Comment (VineBalloon) [inline config/default.ini]: Include useful comments on where to find these keys. Perhaps also create rules in the configparser that automatically switch off commands that rely on the keys if the aren't set.

File: bot.py (hunk 1 above)
Comment (TerryAg) [reply bot.py]: All config parser logic is now in a file configstartup.py - I wanted to keep it separate from bot.py

File: bot.py (hunk 2 above)
Comment (TerryAg) [reply bot.py]: Changed comment

File: bot.py (hunk 3 above)
Comment (TerryAg) [reply bot.py]: Exception raised. I don't think it's possible to raise an error with `config.read()` - it just returns False if it was unsuccessful.

File: commands/music.py (hunk 4 above)
Comment (TerryAg) [reply commands/music.py]: Done - code can be found in configstartup.py

File: bot.py [hunk 6]
```diff
@@ -42,28 +46,32 @@
...
         if message.content.startswith(commands.PREFIX) and message.author != client.user:
             args = '\\n '.join(message.content[1:].splitlines()).split()
             if args:
                 cls, args = commands.Helpme.find_command(args)
-                output = await cls(client, message).init(*args)
-                if isinstance(output, discord.Embed):
-                    await client.send_message(message.channel, embed=output)
-                elif output is not None:
-                    await client.send_message(message.channel, output)
+                command = cls.base_command
```
[the remark this answers is not in the document]
Comment (VineBalloon) [reply bot.py]: Ah ok, wasn't familiar with the decorator.

**Pull request merged**